src/plugins/data_views/server/rest_api_routes/internal/fields.ts

```diff
@@ -0,0 +1,140 @@
+  DataViewsServerPluginStart,
+  DataViewsServerPluginStartDependencies,
+} from '../../types';
+import type { FieldDescriptorRestResponse } from '../route_types';
+import { FIELDS_PATH as path } from '../../../common/constants';
+import { parseFields, IBody, IQuery, querySchema } from './fields_for';
+
+export function calculateHash(srcBuffer: Buffer) {
+  const hash = createHash('sha1');
+  hash.update(srcBuffer);
+  return hash.digest('hex');
+}
+
+const handler: (
+  isRollupsEnabled: () => boolean,
+  getUserId: () => (kibanaRequest: KibanaRequest) => Promise<string | undefined>
+) => RequestHandler<{}, IQuery, IBody> =
+  (isRollupsEnabled, getUserIdGetter) => async (context, request, response) => {
+    const core = await context.core;
+    const uiSettings = core.uiSettings.client;
+    const { asCurrentUser } = core.elasticsearch.client;
+    const indexPatterns = new IndexPatternsFetcher(asCurrentUser, undefined, isRollupsEnabled());
+    const {
+      pattern,
+      meta_fields: metaFields,
+      type,
+      rollup_index: rollupIndex,
+      allow_no_index: allowNoIndex,
+      include_unmapped: includeUnmapped,
+    } = request.query;
+
+    let parsedFields: string[] = [];
+    let parsedMetaFields: string[] = [];
+    try {
+      parsedMetaFields = parseFields(metaFields);
+      parsedFields = parseFields(request.query.fields ?? []);
+    } catch (error) {
+      return response.badRequest();
+    }
+
+    try {
+      const { fields, indices } = await indexPatterns.getFieldsForWildcard({
+        pattern,
+        metaFields: parsedMetaFields,
+        type,
+        rollupIndex,
+        fieldCapsOptions: {
+          allow_no_indices: allowNoIndex || false,
+          includeUnmapped,
+        },
+        ...(parsedFields.length > 0 ? { fields: parsedFields } : {}),
+      });
+
+      const body: { fields: FieldDescriptorRestResponse[]; indices: string[] } = {
+        fields,
+        indices,
+      };
+
+      const etag = calculateHash(Buffer.from(JSON.stringify(body)));
+
+      const getUserId = getUserIdGetter();
+      const userId = await getUserId(request);
+      const userHash = userId ? calculateHash(Buffer.from(userId)) : '';
+
+      const headers: Record<string, string> = {
+        'content-type': 'application/json',
+        etag,
+        vary: 'accept-encoding, user-hash',
+        'user-hash': userHash,
+      };
+
+      const cacheMaxAge = await uiSettings.get<number>('data_views:cache_max_age');
+
+      if (cacheMaxAge && fields.length) {
+        const stale = 365 * 24 * 60 * 60 - cacheMaxAge;
+        headers[
+          'cache-control'
+        ] = `private, max-age=${cacheMaxAge}, stale-while-revalidate=${stale}`;
+      } else {
+        headers['cache-control'] = 'private, no-cache';
+      }
+
+      const ifNoneMatch = request.headers['if-none-match'];
+      const ifNoneMatchString = Array.isArray(ifNoneMatch) ? ifNoneMatch[0] : ifNoneMatch;
+
+      if (ifNoneMatchString) {
```

Comment:
```suggestion
      if (ifNoneMatchString && cacheMaxAge > 0) {
```
Shouldn't we just return a "noModified" response, when cache is enabled?

Reply:
Adding this line would disable the potential for 304 NotModified responses when caching is disabled. These are separate caching / optimization tools. With the cache disabled, it still waits for the 304 response and it likely takes the same amount of time. The difference is lack of data transfer.

Reply:
Ah, got it, thx. So in case no fields have changed in the result, this would mean no payload would be submitted, so even if request to ES takes long, the network request from the middleware to the browser would be fast, right

Reply:
correct

Reply:
thx